### PR TITLE
Fix replication_target_status metrics reset and stale series cleanup

### PR DESCRIPTION
**Pull request by aayushchouhan09**
### Describe the Problem
Each scan called `reset()` on replication_target_status but only republished one rule per policy, so multi-rule policies lost metric series every cycle (flapping/absent alerts). Stale series also remained after policy delete or rule changes.

### Explain the Changes
1. Removed per-scan `reset()`
2. bg_workers only — reachability is set to 0/1 per (replication_id, source, target) after each scan; web clears do not affect scraped metrics.
3. Reconcile at scan start — drops stale series for deleted policies, missing source, or outdated rules/labels (no full gauge reset).
4. Helpers — `set_replication_target_status` and `clear_replication_target_status_by_replication_id` in noobaa_core_report, used via replication_utils (update, clear_orphan, reconcile).

### Issues: Fixed #xxx / Gap #xxx
1. Fixed: https://redhat.atlassian.net/browse/DFBUGS-6398
2. Gap: #9593 

### Testing Instructions:
1.  To run unit tests: `npx jest src/test/unit_tests/internal/test_replication_metrics.test.js`

Manual testing steps:
1. Create 4 destination/target buckets obcs dst-n,dst-s,dst-w and dst-e.
2. Create two policies policy1.json and policy2.json, and put two buckets in each replication policy.

> $ cat policy.json 
{
  "rules": [
    {
      "rule_id": "rule-1",
      "destination_bucket": "<dst-n-name>"
    },
    {
      "rule_id": "rule-2",
      "destination_bucket": "<dst-s-name>"
    }
  ]
}

3. Create two src buckets obcs src-ns and src-we by using above policies.
4. Verify if replication is working fine, and also check the replication_target_status metrics using below commands (It should contain 4 metrics for each replication rule_id having reachable state).                                                                          
`$ JWT_TOKEN=$(kubectl get secret/noobaa-metrics-auth-secret -o jsonpath={.data.metrics_token} | base64 -d)`
`$ kubectl exec -it noobaa-core-0 -- curl -k -H "Authorization: Bearer ${JWT_TOKEN}" http://localhost:8080/metrics/bg_workers | grep replication_target_status`
5. Try deleting 1 destination bucket, and check the metrics (it will take time to update as replication cycle execute one rule in every `5 min~can_be_changed` by default) --> for that particular src and dst bucket flag should be unreachable/0
6. Now, try deleting replication on src bucket or delete the any of the source bucket, and check the metrics --> all the rows should be removed container that replication_id or src bucket
7. You can try different combinations and check if the metrics are reflecting correct values.

- [ ] Doc added/updated
- [X] Tests added

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved cleanup of replication target status for orphaned policies when source buckets are missing.
  * More accurate reachability updates and failure reporting during replication scans.
  * Metric operations now safely no-op when metrics are unavailable.

* **Refactor**
  * Replaced bulk-reset with a targeted reconciliation flow that prunes stale replication metrics and validates against current policies.

* **Tests**
  * Added unit tests covering reconciliation, cleanup, deleted/dangling buckets, and edge cases in metric handling.

[Review Change Stack](https://app.coderabbit.ai/change-stack/noobaa/noobaa-core/pull/9623?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by shirady** [comment]
@aayushchouhan09 can you update the PR description?
Can you add manual testing instructions? (in a comment or in the PR description)

File: src/server/analytic_services/prometheus_reports/noobaa_core_report.js [hunk 1]
```diff
@@ -705,15 +705,20 @@ class NooBaaCoreReport extends BasePrometheusReport {
...
-        this._metrics.replication_target_status.reset();
+        this._metrics.replication_target_status.set({ replication_id, source_bucket, target_bucket }, Number(is_reachable));
     }
 
-    set_replication_target_status(replication_id, source_bucket, target_bucket, is_reachable) {
+    clear_replication_target_status_by_replication_id(replication_id) {
         if (!this._metrics) return;
-        this._metrics.replication_target_status.set({ replication_id, source_bucket, target_bucket }, Number(is_reachable));
+        const id = String(replication_id);
+        const gauge = this._metrics.replication_target_status;
```
Comment (shirady) [inline src/server/analytic_services/prometheus_reports/noobaa_core_report.js]: What ensures that `gauge.hashMap` is not undefined?

File: src/server/analytic_services/prometheus_reports/noobaa_core_report.js [hunk 2]
```diff
@@ -705,15 +705,20 @@ class NooBaaCoreReport extends BasePrometheusReport {
...
+        for (const key of Object.keys(gauge.hashMap)) {
+            if (gauge.hashMap[key]?.labels?.replication_id === id) delete gauge.hashMap[key];
+        }
```
Comment (shirady) [inline src/server/analytic_services/prometheus_reports/noobaa_core_report.js]: Could you add a short comment about this loop?
If you think it is too obvious, then - no, just explain here (PR).

File: src/server/bg_services/replication_scanner.js [hunk 3]
```diff
@@ -72,8 +71,11 @@ class ReplicationScanner {
             const { src_bucket, dst_bucket } = replication_utils.find_src_and_dst_buckets(rule.destination_bucket, replication_id);
             if (!src_bucket || !dst_bucket) {
                 dbg.error('replication_scanner: can not find src_bucket or dst_bucket object', src_bucket, dst_bucket);
-                // mark target bucket as unreachable (resolve display name from id while bucket row still exists)
-                if (src_bucket && !dst_bucket) {
+                if (!src_bucket) {
+                    replication_utils.clear_replication_target_status_for_orphan_policy(replication_id);
+                    return;
+                }
+                if (!dst_bucket) {
```
Comment (shirady) [inline src/server/bg_services/replication_scanner.js:78]: If we don't have `dst_bucket` don't we need to call `clear_replication_target_status_for_orphan_policy`? (like the condition `if (!src_bucket)` above).

File: src/test/unit_tests/internal/test_replication_metrics.test.js [hunk 4]
```diff
@@ -0,0 +1,201 @@
+/* Copyright (C) 2016 NooBaa */
```
Comment (shirady) [inline src/test/unit_tests/internal/test_replication_metrics.test.js]: New file, you can update the year.

File: src/test/unit_tests/internal/test_replication_metrics.test.js [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+'use strict';
+
+// bg_workers gauge: reconcile fixes stale series after policy delete, src removal,
+// rule change, id-vs-name labels, and -deleting-* names (web clear does not apply here).
```
Comment (shirady) [inline src/test/unit_tests/internal/test_replication_metrics.test.js]: Do you need this comment or is it redundant?

File: src/test/unit_tests/internal/test_replication_metrics.test.js [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+        mock_store.deleted.clear();
+        mock_store.get_by_id = id => mock_store.buckets.find(doc => String(doc._id) === String(id));
+        mock_get_replication_by_id.mockReset();
+        mock_core.mockReturnValue(report);
+    });
+
+    it.each([
+        [
+            'removes all stale metrics when replication policy was deleted from DB (mark_deleted)',
+            () => mock_get_replication_by_id.mockResolvedValue(null),
```
Comment (shirady) [inline src/test/unit_tests/internal/test_replication_metrics.test.js]: Do you get `null` in real replication by ID?

File: src/test/unit_tests/internal/test_replication_metrics.test.js [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        report._metrics = null;
+        report.set_replication_target_status(R1, SRC, DST, 1);
+        report.clear_replication_target_status_by_replication_id(R1);
+        report._metrics = m;
+        expect(Object.keys(g().hashMap)).toHaveLength(0);
+    });
+
+    it.each([
+        [
+            'returns bucket name from soft-deleted row via get_by_id_include_deleted',
```
Comment (shirady) [inline src/test/unit_tests/internal/test_replication_metrics.test.js]: "soft-deleted row" I'm not sure I understand it

File: src/test/unit_tests/internal/test_replication_metrics.test.js [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+            [DST],
+        ],
+    ])('reconcile at scan start: %s', async (_desc, setup, want) => {
+        probe(R1, SRC, DST);
+        setup();
+        await reconcile();
+        expect(tgts(R1)).toEqual(want);
+    });
+
+    it('reconcile prunes targets removed from policy and drops series where target_bucket is destination object id', async () => {
```
Comment (shirady) [inline src/test/unit_tests/internal/test_replication_metrics.test.js]: I'm not sure I understand this title

File: src/test/unit_tests/internal/test_replication_metrics.test.js [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+        probe(R1, SRC, 'stale');
+        probe(R1, SRC, 'dst-id-a');
+        g().hashMap.bad = { labels: null, value: 0 };
+        seed(R1, [DST, 'dst-b']);
+        await reconcile();
+        expect(tgts(R1).sort()).toEqual([DST, 'dst-b'].sort());
+        expect(g().hashMap.bad).toBeDefined();
+    });
+
+    it('clear_replication_target_status_for_orphan_policy clears one policy only; null replication_id is ignored', () => {
```
Comment (shirady) [inline src/test/unit_tests/internal/test_replication_metrics.test.js]: As already asked, I'm not sure you would have a `null` value, only `undefined`.

File: src/test/unit_tests/internal/test_replication_metrics.test.js [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+        ],
+    ])('reconcile early exit: %s', async (_desc, prep) => {
+        prep();
+        await reconcile();
+        expect(mock_get_replication_by_id).not.toHaveBeenCalled();
+    });
+
+    it('NooBaaCoreReport set_replication_target_status and clear do nothing when _metrics is null', () => {
+        const m = report._metrics;
+        report._metrics = null;
```
Comment (shirady) [inline src/test/unit_tests/internal/test_replication_metrics.test.js:163]: Is `report._metrics` `null` or just `undefined`?

File: src/test/unit_tests/internal/test_replication_metrics.test.js [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+
+    it.each([
+        [
+            'returns bucket name from soft-deleted row via get_by_id_include_deleted',
+            () => mock_store.deleted.set('d', b({ _id: 'd', name: DST, deleting: new Date() })),
+            'd',
+            DST,
+        ],
+        [
+            'falls back to bucket id when live and deleted rows are missing',
```
Comment (shirady) [inline src/test/unit_tests/internal/test_replication_metrics.test.js]: Could you improve the title?
"falls back to bucket id when live and deleted rows are missing"
What is "when live"?

File: src/test/unit_tests/internal/test_replication_metrics.test.js [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+            () => undefined,
+            'gone',
+            'gone',
+        ],
+        [
+            'falls back to bucket id when system_store lookup throws',
```
Comment (shirady) [inline src/test/unit_tests/internal/test_replication_metrics.test.js]: Could you improve the title?
```suggestion
            'falls back to bucket id when system_store lookup throws an error',
```
What is the context of the fallback?

File: src/test/unit_tests/internal/test_replication_metrics.test.js [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+            'd',
+            '',
+        ],
+        [
+            'returns empty string when deleting bucket has empty name before suffix strip',
+            () => { mock_store.buckets = [b({ _id: 'd', name: '', deleting: new Date() })]; },
+            'd',
+            '',
+        ],
+        [
```
Comment (shirady) [inline src/test/unit_tests/internal/test_replication_metrics.test.js]: > returns empty string when bucket name unwraps to empty'
> 'returns empty string when deleting bucket has empty name before suffix strip'

Do you have cases where the bucket name is empty?
Is the mock directly on the DB? I'm not sure it can happen

File: src/server/utils/replication_utils.js [hunk 14]
```diff
@@ -106,17 +107,63 @@ function report_failed_replication_cycle(bucket_name, replication_id, rule_id, t
...
 function update_replication_target_status(replication_id, source_bucket, target_bucket, is_reachable) {
     if (!replication_id) return;
     const core_report = prom_reporting.get_core_report();
     const src_name = source_bucket instanceof SensitiveString ? source_bucket.unwrap() : source_bucket;
     const dst_name = target_bucket instanceof SensitiveString ? target_bucket.unwrap() : target_bucket;
-    core_report.set_replication_target_status(replication_id, src_name, dst_name, is_reachable);
+    core_report.set_replication_target_status(String(replication_id), src_name, dst_name, is_reachable);
+}
+
+function clear_replication_target_status_for_orphan_policy(replication_id) {
```
Comment (shirady) [inline src/server/utils/replication_utils.js:120]: Could you add a comment what " orphan_policy " exactly is?

File: src/server/utils/replication_utils.js [hunk 15]
```diff
@@ -106,17 +107,63 @@ function report_failed_replication_cycle(bucket_name, replication_id, rule_id, t
...
+    if (!replication_id) return;
+    const core_report = prom_reporting.get_core_report();
+    core_report.clear_replication_target_status_by_replication_id(String(replication_id));
+}
+
+// bg_workers: prune stale series for replication_ids that already have metrics
```
Comment (shirady) [inline src/server/utils/replication_utils.js]: Is it defined somewhere what "stale " is?

**Review by shirady — COMMENTED** [review]
I tried to raise questions and added a few minor comments

File: src/server/analytic_services/prometheus_reports/noobaa_core_report.js (hunk 2 above)
Comment (aayushchouhan09) [reply src/server/analytic_services/prometheus_reports/noobaa_core_report.js]: We are deleting all the metrics entries here on the basis of `replication_id` in the argument.

File: src/server/bg_services/replication_scanner.js (hunk 3 above)
Comment (aayushchouhan09) [reply src/server/bg_services/replication_scanner.js:78]: No, in case of `dst_bucket` is not found we are updating the flag to unreachable, we don't need to delete the entries. And with this entries we are firing the alert for target/dst bucket.

If there is no src bucket present or replication is deleted then the policy are of no use and we are removing there entries only.

File: src/server/utils/replication_utils.js (hunk 15 above)
Comment (aayushchouhan09) [reply src/server/utils/replication_utils.js]: Updated definition, Thanks

File: src/test/unit_tests/internal/test_replication_metrics.test.js (hunk 6 above)
Comment (aayushchouhan09) [reply src/test/unit_tests/internal/test_replication_metrics.test.js]: yes it returns `null` or `undefined` if no rules found in DB

File: src/test/unit_tests/internal/test_replication_metrics.test.js (hunk 10 above)
Comment (aayushchouhan09) [reply src/test/unit_tests/internal/test_replication_metrics.test.js:163]: NooBaaCoreReport sets this._metrics = null when PROMETHEUS_ENABLED is false.

Updated title, Thanks

File: src/test/unit_tests/internal/test_replication_metrics.test.js (hunk 7 above)
Comment (aayushchouhan09) [reply src/test/unit_tests/internal/test_replication_metrics.test.js]: If a bucket is deleted from the cluster, its entry still present in the DB (for upto 3months) until the DB cleaner removes it. And the bucket name is suffixed with `-deleting-<timestamp>` in that case and these rows are being called as soft-deleted rows in DB.
We can still fetch the name of the deleted bucket from that row.

File: src/test/unit_tests/internal/test_replication_metrics.test.js (hunk 11 above)
Comment (aayushchouhan09) [reply src/test/unit_tests/internal/test_replication_metrics.test.js]: When we are not able to fetch the bucket name even from the soft-deleted rows, we fallback and print the bucket_id instead of name in the alert.

File: src/test/unit_tests/internal/test_replication_metrics.test.js (hunk 12 above)
Comment (aayushchouhan09) [reply src/test/unit_tests/internal/test_replication_metrics.test.js]: When we are not able to fetch the bucket name even from the soft-deleted rows, we fallback and print the bucket_id instead of name in the alert.

File: src/test/unit_tests/internal/test_replication_metrics.test.js (hunk 13 above)
Comment (aayushchouhan09) [reply src/test/unit_tests/internal/test_replication_metrics.test.js]: We cannot have such cases, removing the two test cases. Thanks

File: src/server/analytic_services/prometheus_reports/noobaa_core_report.js [hunk 16]
```diff
@@ -705,15 +705,22 @@ class NooBaaCoreReport extends BasePrometheusReport {
...
     set_replication_target_status(replication_id, source_bucket, target_bucket, is_reachable) {
         if (!this._metrics) return;
         this._metrics.replication_target_status.set({ replication_id, source_bucket, target_bucket }, Number(is_reachable));
     }
+
+    clear_replication_target_status_by_replication_id(replication_id) {
+        const gauge_hash_map = this._metrics?.replication_target_status?.hashMap;
+        if (!gauge_hash_map) return;
+
+        const id = String(replication_id);
```
Comment (shirady) [inline src/server/analytic_services/prometheus_reports/noobaa_core_report.js:717]: Do you guard somewhere that `replication_id` is not `undefined`?

File: src/server/utils/replication_utils.js [hunk 17]
```diff
@@ -106,17 +107,66 @@ function report_failed_replication_cycle(bucket_name, replication_id, rule_id, t
...
+    core_report.clear_replication_target_status_by_replication_id(String(replication_id));
+}
+
+// bg_workers: reconcile target reachability metrics at replication scan start
+// compare exported series to replication DB: clear orphans, remove source/target labels that no longer match
+async function reconcile_replication_target_status() {
+    const gauge = prom_reporting.get_core_report()._metrics?.replication_target_status;
+    if (!gauge?.hashMap) return;
+
+    // collect replication_ids that already have exported series
```
Comment (shirady) [inline src/server/utils/replication_utils.js]: "exported series"?
I'm not I understand it, if you can define what it is in brackets.

File: src/test/unit_tests/internal/test_replication_metrics.test.js (hunk 11 above)
Comment (shirady) [reply src/test/unit_tests/internal/test_replication_metrics.test.js]: Could you add an example to show it?

**Review by shirady — APPROVED** [review]
Approving not to block you - added final comments

File: src/test/unit_tests/internal/test_replication_metrics.test.js (hunk 11 above)
Comment (aayushchouhan09) [reply src/test/unit_tests/internal/test_replication_metrics.test.js]: We have added a fix for it in this [PR](https://github.com/noobaa/noobaa-core/pull/9566/changes) where we resolving the bucket name, you can see the testing steps there.
This is the worst case scenario when soft-deleted rows are also deleted from the DB by the `DB_cleaner`

File: src/server/utils/replication_utils.js [hunk 18]
```diff
@@ -106,17 +107,66 @@ function report_failed_replication_cycle(bucket_name, replication_id, rule_id, t
...
+        const src_name = bucket_name_for_metrics(src_bucket);
+
+        // build label set that should exist per current DB rules
+        const valid_keys = new Set();
+        for (const rule of rules) {
+            const dst_name = await resolve_destination_bucket_name(rule.destination_bucket);
+            valid_keys.add(`${repl_id}|${src_name}|${dst_name}`);
+        }
+        // remove (src, dst) series that no longer match DB rules for this replication_id
+        for (const key of Object.keys(gauge.hashMap)) {
```
Comment (naveenpaul1) [inline src/server/utils/replication_utils.js]: going through complete metrics and removing one could be resourse consuming, If possible refactor this, But shouldnt be a blocker for PR Merge, LGTM

**Review by naveenpaul1 — APPROVED** [review]

File: src/server/analytic_services/prometheus_reports/noobaa_core_report.js (hunk 16 above)
Comment (aayushchouhan09) [reply src/server/analytic_services/prometheus_reports/noobaa_core_report.js:717]: We are guarding it from replication_utils caller functions for both update and clear.

File: src/server/utils/replication_utils.js (hunk 18 above)
Comment (aayushchouhan09) [reply src/server/utils/replication_utils.js]: I did the refactor and calling DB once to fetch all the rules and then clearing the metrics using one loop. Please notice here `find_src_and_dst_buckets()` is not quering the DB.
I think this makes it memory efficient. Thanks!
